js/app/test/cancel_events.spec.ts

```diff
@@ -0,0 +1,40 @@
+import { test, expect } from "@gradio/tootils";
+
+test("yielding", async ({ page }) => {
+	let first_iteration;
+	let fourth_iteration;
+	let last_iteration;
+
+	page.on("websocket", (ws) => {
+		first_iteration = ws.waitForEvent("framereceived", {
+			predicate: (event) => {
+				return JSON.parse(event.payload as string).msg === "process_generating";
+			}
+		});
+
+		fourth_iteration = ws.waitForEvent("framereceived", {
+			predicate: (event) => {
+				return JSON.parse(event.payload as string)?.output?.data?.[0] === "3";
+			}
+		});
+
+		last_iteration = ws.waitForEvent("framereceived", {
+			predicate: (event) => {
+				return JSON.parse(event.payload as string).msg === "process_completed";
+			}
+		});
```

Comment:
Here we are using `waitForEvent` with a predicate function. The predicate dictates whether or not we have received the event we are waiting for. We are checking against the actual content of the message here to make sure it is the correct one.

`waitForEvent` returns a promise, we don't actual await it here because we don't want to block anything just yet. We want to interweave our 'blocks' with assertions later on. Promises in javascript will start/continue in the background if you do not await them directly. So instead of awaiting here, we assign these promises the references we defined earlier so we can wait for them later.